pkg/indexed_blob/lib/indexed_blob.dart

```diff
@@ -0,0 +1,326 @@
+  /// is not allowed.
+  ///
+  /// The [blobId] is a free-form string that is included when a FileRange is
+  /// returend by [BlobIndex.lookup]. It is intended to store an identifier for
+  /// blob containing the file, or a URL to the blob.
+  Future<BlobIndex> buildIndex(String blobId) async {
+    _checkState();
+    _finished = true;
+    await _blob.close();
+
+    final out = StringBuffer();
+    final w = jsonStringWriter(out);
+    w.startObject();
+
+    // Note. it's a bit of a hack, but keys are expected to be order as follows:
+    // - version
+    // - blobId
+    // - index
+
+    // Add version number
+    w.addKey('version');
+    w.addNumber(1);
+
+    // Add blobId
+    w.addKey('blobId');
+    w.addString(blobId);
+
+    // Add index
+    w.addKey('index');
+    _addMap(w, _index);
+
+    w.endObject();
+
+    final bytes = utf8.encoder.convert(out.toString());
+
+    return BlobIndex.fromBytes(bytes);
+  }
+}
+
+void _addMap(JsonSink w, Map<String, dynamic> m) {
+  w.startObject();
+  for (final kv in m.entries) {
+    w.addKey(kv.key);
+
+    final value = kv.value;
+    if (value is String) {
+      w.addString(value);
+    } else if (value is Map<String, dynamic>) {
+      _addMap(w, value);
+    } else {
+      throw AssertionError('unexpected value: $value');
+    }
+  }
+  w.endObject();
+}
+
+bool _skipUntilKey(JsonReader r, String key) {
+  while (r.tryKey([key]) == null) {
+    if (!r.skipObjectEntry()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Index for an indexed-blob, can be used to lookup [FileRange].
+///
+/// The internal byte format of a [BlobIndex] is encoded as JSON, but the order
+/// of properties is significant and the format should be considered internal
+/// to [BlobIndex] and [IndexedBlobBuilder].
+///
+/// ## Internal BlobIndex Format v1.0
+///  * This format is **internal** to [BlobIndex] and [IndexedBlobBuilder].
+///  * Order of properties is significant,
+///  * Additional top-level properties must be ignored.
+///
+/// ```json
+/// {
+///   // First property must be version, the version is bumped when backwards
+///   // compatibility is broken. Additional meta-data can be introduced without
+///   // breaking compatibility as additional top-level properties **must** be
+///   // ignored.
+///   "version": 1,
+///
+///   // Identifier of the blob this index is pointing into.
+///   "blobId": "<blobId>",
+///
+///   // Index into the blob, with paths split at '/' each folder becomes a map.
+///   "index": {
+///     // ordering within "index" property is insignificant in version 1.x
+///     "README.md": "80:100",
+///     "bin": {
+///       "pub_worker.dart": "0:42"
+///     },
+///     "lib": {
+///       // File "lib/blob.dart" starts at 42 and ends at 80 in <blobId>.
+///       "blob.dart": "42:80"
+///     }
+///   }
+/// }
+/// ```
+///
+@sealed
+class BlobIndex {
+  final Uint8List _indexFile;
```

Comment:
Note: I'm a bit worried that each path access is another partial parsing of the json, but maybe that's all we need. It is hard to tell how efficient this will be, let's go with it, and test it in practice. We have at least one alternative format that can be sharded and we could also try out if needed.